### DISCUSSION TITLE
Feature Request: Automatic Prompt Caching Support for Anthropic API (Please 🙏🏻)

### DISCUSSION THREAD

**Discussion by IN-Neil** [open-webui/open-webui#7873] [Ideas]
I recently switched from LibreChat to Open WebUI and noticed the API costs blew up, particularly for long conversations (22k+ tokens that that common for me). I didn't realize how useful this was until I switched (LibreChat automatically handled prompt caching with Anthropic's API). 

Current Situation:
- There's a pipe that includes capability for prompt caching but it doesn't automatically structure prompts to use cache_control
- Long conversations a lot more expensive without caching, $10, gone in few hours, vs days with caching. 

Request:
1. Add automatic prompt caching (option toggle) like LibreChat does
2. Implement automatic cache_control structuring for more static stuff like:
   - System prompts, 
   - Repetitive contexts, docs

Benefits:
- Up to 90% cost reduction for cached content
- Better performance for long conversations
- Improved user experience for Claude users migrating from LibreChat

Would it be possible to add this functionality in any way? Let me know if I'm missing anything. I head OpenRouter can do this, but it's not an option for me right now.

**Comment by Vandit13S** [comment]
Much required!
BTW what's the exact problem you are facing with the pipe?

**Comment by crizCraig** [comment]
Related: https://github.com/open-webui/open-webui/issues/4887 

I'm using prompt caching in my hosted OpenWeb UI, [PolyChat](https://polychat.co), and it's reduced costs by 66%. Here's the [code](https://gist.github.com/crizCraig/6f8885e8cd9ecaf1ee333b0ec3c30de0) for it. Since I implemented this in the API layer, not part of OpenWeb UI, it's not ready as a PR, but it's a self-contained module with tests.

I set the cache_threshold and max_context_chars to 8k chars (~2k tokens) and 800k chars (~200k tokens) respectively since Sonnet 3.5 token limit is 200k.

**Reply by jojje** [reply]
I guess this can be considered addressed now, since there are a couple of "functions" on the [function](https://openwebui.com/functions) page that do caching already.

Just mentioning it here, since this was the first hit I got when I had the same question and need; to save money through caching. So if anyone else lands here for that same goal, above is the link to remedy.

**Reply by crizCraig** [reply]
Nice, would be nice to get it to default enabled without needing to use a function, but I've found [this](https://github.com/open-webui/open-webui/issues/4887#issuecomment-2599335217) edge case where you don't want to cache RAG prompts, which I'm not sure the functions handle. I updated my [code](https://gist.github.com/crizCraig/6f8885e8cd9ecaf1ee333b0ec3c30de0) with a bit of a hack to check for it, but if implemented in `main` it would be better to check the request.type for `retrieval`.

**Reply by DrewWalkup** [reply]
@crizCraig How would one implement your script into their own open webui system?

**Comment by darkBuddha** [comment]
How can such an important feature not be supported

**Comment by Podden** [comment]
I tried to Implement promt caching as best as possible in my anthropic pipe. It detects if the request uses knowledge or files with RAG or the memory System and extracts this from the system promt to the last user Message and moves the Cache breakpoint to the previous Message. Give it a try, you can enable usage statistics via uservalve.

https://openwebui.com/posts/complex_anthropic_pipe_d51d4333

Keep in mind that System promt or Tool change always invalidates the full message cache so using things like Auto tool Filter or an often changing dynamic variable like %time% is not an option.

**Comment by pomology** [comment]
Is this anticipated as an addition? It would sure be useful.
